bindings/rust-examples/client-hello-config-resolution/src/bin/async_load_server.rs

```diff
@@ -0,0 +1,136 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+use s2n_tls::{
+    callbacks::{ClientHelloCallback, ConfigResolver, ConnectionFuture},
+    config::Config,
+    security::{Policy, DEFAULT_TLS13},
+};
+use s2n_tls_tokio::TlsAcceptor;
+use std::{error::Error, future::Future, pin::Pin};
+use tokio::{io::AsyncWriteExt, net::*, try_join};
+
+const PORT: u16 = 1738;
+
+#[derive(Clone)]
+pub struct AsyncAnimalConfigResolver {
+    // the directory that contains the relevant certs
+    cert_directory: String,
+}
+
+struct SpecificAnimalResolver {
+    cert_directory: String,
+    animal: String,
+}
+
+impl AsyncAnimalConfigResolver {
+    fn new(cert_directory: String) -> Self {
+        AsyncAnimalConfigResolver { cert_directory }
+    }
+
+    // This method will lookup the appropriate certificates and read them from disk
+    // in an async manner which won't block the tokio task.
+    //
+    // Note that this method consumes `self`. A ConfigResolver can be constructed
+    // from a future that returns `Result<Config, s2n_tls::error::Error>`, with
+    // the main additional requirements that the future is `'static`. This generally
+    // means that it can't have any  interior references.
+    //
+    // If this method took in `&self`, then
+    // ```
+    // let config_resolver = ConfigResolver::new(self.server_config(animal));
+    // ```
+    // wouldn't compile because the compiler would complain that `&self` doesn't
+    // live long enough.
```

Comment:
Could you maybe solve this more intuitively by returning a clone of config_resolver instead of consuming a clone of async_resolver_clone? Or does that not work?

Reply:
I'm not sure that I'm following (I need to edit that doc comment to refer to the explicit `AsyncAnimalConfigResolver`, which might have made things more confusing)

`server_config(animal)` could return a type which directly implements the `Future` trait, but that would require customers to write all of the actual polling logic for the type, which is very complex and unpleasant.

I originally had this structure as taking in `&self` but that was also a little tricky, see: https://github.com/aws/s2n-tls/pull/4477#discussion_r1612221593

I agree that this is a bit unpleasant to deal with, but I think returning a type that directly implemented `Future` would be even less pleasant ☹️